7주차 과제 - 라스

MagicalLas opened the pull request.

MagicalLas committed:
이번주에 집중 할 것들을 적어보았어요

hannut91 commented:
목표가 확실하시군요 ㅎㅎ 이번주 화이팅입니다!

MagicalLas commented:
이번주는 너무 힘드네요. 천천히 차근차근 해나가겠습니다. 저는 쓰레기를 낼꺼에요.

일단 가장 처음 할 일은 lombok을 제거하면서 user를 리팩터링 하는거에요.
저는 lombok이 너무 고통스럽습니다. Builder도 마찬가지이구요. User의 changeWith도 고통스러워요.
그러니 일단 User를 POJO로 만들면서 리팩터링을 해볼게요.

MagicalLas committed:
Refactor User's changeWith method

유저 객체의 changeWith메서드를 리팩터링 하였습니다.
유저 객체를 변경하기 위하여 유저 객체를 받는 것은 불편한 것 같습니다.
장기적으로는 Lombok을 User에서 걷어내는 것이 목표입니다.
Mapper도 같이 걷어낼 것 같아요.

MagicalLas commented:
User에서 Lombok을 없애려고 하니까 테스트가 매우 많이 깨지네요. 기존에 Lombok에 너무 종속적인 코드였던 것 같아요. 특정 라이브러리에 이렇게 의존하는건 안좋은 것 같아요. (그래서 annotation이 없는 POJO를 좋아합니다.)

MagicalLas committed:
Remove lombok from user model

Lombok을 User에서 제거하였습니다.
드디어 유저가 깨끗해졌네요!

MagicalLas committed:
Add User object default constructor for Hibernate

기존에 User에서는 Lombok의 @NoArgsConstructor을 이용해서 default constructor를 자동으로 정의해주고 있었습니다.

하지만, 098246159b62910ecfcef0a8a387c70b3922b7db 커밋에서 User도메인 객체의 Lombok을 없애면서 user객체의 default constructor가 정의되지 않았습니다.

default constructor는 Hibernate에서 사용하고 있기 때문에 정의를 해주어야 합니다.

따라서, 유저 객체의 default constructor를 직접 정의해주었습니다.

See Also
  - https://docs.jboss.org/hibernate/orm/3.5/reference/en/html/persistent-classes.html#persistent-classes-pojo-constructor

hannut91 commented:
외부 라이브러리에 의존하는 것은 저도 좋아하지 않습니다 ㅎㅎ 하지만 너무 많은 코드를 작성해야하는 번거로움 때문에 Lombok을 쓸 수 밖에 없게 되는 것 같아서 저도 참 아쉽습니다.

코틀린이 그립습니다...

MagicalLas commented:
코틀린 쓰고싶네요...
오늘도 힘내서 진행할 예정이지만, 조금 늦게 올라올 것 같아요. 회사에서 좀 안좋은 일이 있어서 이제 집에 도착했네요.
주말내내 천천히 발전해나갈 예정이니 여유로운 마음으로 커밋 하나하나 봐주시면 좋을 것 같아요!

hannut91 commented:
기대가 되네요 ㅎㅎㅎ

hannut91 commented:
기대가 되네요 ㅎㅎㅎ

MagicalLas commented:
User클래스에서 Autowired를 사용해서 EncryptService를 쓰려고 하는데 어려움이 있네요. Spring의 bean을 테스트 환경에서 어떻게 넣어줄 수 있는지 모르겠어요.
열심히 해봤는데, 잘 안되네요ㅠㅠ
왜 유저에 안들어가니...

MagicalLas committed:
유저 비밀번호 암호화 구현

기존에는 유저의 비밀번호를 평문으로 저장하고 있었습니다.
하지만, 유저의 비밀번호는 보안상 암호화 하여 저장해야 합니다.
따라서, 유저의 비밀번호를 암호화 해서 저장하도록 수정하였습니다.

그 과정에서 암호화 하는 부분을 서비스로 분리하여 구현하였습니다.
의존성 주입을 도와주기 위하여 Container를 만들었으며, 의존성 주입을 가능하도록 만들었습니다.

MagicalLas commented:
휴 겨우겨우 했네요. Spring을 사용하려다가 고생했네요. 다른 언어에서 하는 것 처럼 구현했고, Spring도 조금 이용했어요.

MagicalLas committed:
Spring 관련한 코드를 web package로 이동

기존에는 controller, intercepter, config등등 많은 spring관련한 코드가 com.codesoom.assignment 패키지에 나열되어 있었다.
하지만, Clean Architecture에서는 프로젝트를 보면 Spring인지보다는 어떤 일을 하는지 소리쳐야한다고 한다.
따라서, 스프링에 관련된 코드를 com.codesoom.assignment.web 패키지 아래로 옮긴다.

See Also
  - https://blog.cleancoder.com/uncle-bob/2011/09/30/Screaming-Architecture.html

MagicalLas commented:
기존 코드를 너무 바꾸는 걸까 고민이 되네요.

MagicalLas commented:
이제 role을 추가해보겠어요!

MagicalLas commented:
IDDD를 보고있는데, Role이 Group이라는 Aggregate을 갖고있네요. Aggregate이 다른 Aggregate을 갖고있는건 안좋다고, VO인 ID로 통합해야한다고 적혀있던것 같은데...

MagicalLas committed:
Remove unused mapper

MagicalLas commented:
mapper를 안쓰는데 주입받고있어서 지웠어요. 작성하는 코드 수가 줄어들 수는 있지만, 저는 mapper를 직접 작성해도 괜찮을 것 같다는 생각이 들어요. mapper를 쓰는 이유는 changeWith에서 Entity를 받기 때문인 것 같은데, changeWith라는 메서드가 너무 일반적인 메서드라서 별로 좋지 않다는 생각이 드네요.
현재는 복잡한 도메인을 코딩하고 있는게 아니라서 간단한 메서드를 유지해도 괜찮다는 느낌은 들지만, 여전히 저는 어색하네요.

MagicalLas committed:
Remove lombok from user dto

유저 DTO들에게서 Lombok을 제거하였습니다.
꼭 Lombok을 사용하지 않아도 되는 부분이었기에 지우고 직접 getter를 구현했습니다.

Related Commit: 098246159b62910ecfcef0a8a387c70b3922b7db

MagicalLas commented:
너무 Old하게 개발하고 있는 것 같아서 걱정이 되네요. annotation도 힘들어하고, Builder도 싫어하고.. 모던한 자바는 어떤걸까요

이번 과제를 잘 할지 잘 모르겠어서 하고있는 일들을 계속 돌아보게 되네요. 올바르게 가고있는지 계속 확인하게 되는 것 같아요.

hannut91 commented:
확실히 제가 기대한 모양이랑은 다른 것 같네요. 더 나은 방향이 어떤 방향인지도 사실 잘 모르겠습니다.

우리가 코드를 잘 작성해야 하는 이유는 변경에 대한 비용을 줄이기 위해서잖아요. 그런데 우리가 Lombok을 사실 다른 것으로 대체 할 일은 거의 없을 것 같기도하고, 그렇다고 해서 라이브러리에 무작정 의존하는 것도 싫고 고민이 되는 시점인 것 같습니다.

다른 사람들이 현재 어떻게 개발하고 있는지도 많이 눈여겨 봐야 할 것 같습니다. 근데 이러려면 다른 사람들의 코드를 많이 보고, 내가 직접 만들어보고 하는 수 밖에는 없는 것 같아요.

그렇다고 다른 사람들이 이렇게 하니까 무작정 따라하기는 또 싫고, 그러려면 라스님이 지금 하고 계신 것 처럼 IDDD에서 배운 것들을 적용해보야 하고, 계속 시행착오를 겪고 있는 것 같습니다

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -28,15 +26,21 @@ public User registerUser(UserRegistrationData registrationData) {
         }
 
-        User user = mapper.map(registrationData, User.class);
+        User user = new User(
+            null,
+            email,
+            registrationData.getName(),
+            registrationData.getPassword());
```
`null`을 넘겨줘야 하는군요. null을 넘겨주는 것은 `User`를 사용하는 쪽에서 다시 `null`인지 아닌지에 따라서 처리해야 된다는 것을 암시합니다. null을 사용하지 않으려면 어떻게 해야할까요

MagicalLas replied:
기존에 User를 Mapper로 만들어줄때도 null을 넣어주고있었습니다.
지금 생각나는건 2가지정도가 있는데, Repository에서 nextId를 만들어서 생성시점에 넣어주는 것이 있고, 다른 하나는 EmptyId같은 Id를 하나 만들어서 명시적으로 Id가 설정되지 않았다고 하는 방법이 있겠네요.
Jpa를 조금 더 살펴봐야 할 것 같습니다.

null을 넘겨주는것이 안좋다는걸 우리는 알고있지만 들어나지 않았던 이유는 mapper로 숨겨져 있기 때문이라고 생각해요.
기존에 숨겨져있는 것 보다는 지금이 더 명시적으로 null을 주고있는 것으로 보이는데, 저는 숨겨져 있는 문제보다 문제가 들어나는게 좋은 것 같아요.

MagicalLas commented:
> 우리가 코드를 잘 작성해야 하는 이유는 변경에 대한 비용을 줄이기 위해서잖아요. 그런데 우리가 Lombok을 사실 다른 것으로 대체 할 일은 거의 없을 것 같기도하고, 그렇다고 해서 라이브러리에 무작정 의존하는 것도 싫고 고민이 되는 시점인 것 같습니다.

처음부터 Lombok을 쓰지 않는다는 선택을 했다면 좋았을 것 같네요. Lombok을 사용하면서 제가 불편함을 느끼고 개발 생산성이 떨어진다면 Lombok을 없애는게 기술부채를 해결하는 방법이 아니었을까요?

> 확실히 제가 기대한 모양이랑은 다른 것 같네요. 더 나은 방향이 어떤 방향인지도 사실 잘 모르겠습니다.

기대하신 모양은 어떤 모양이었을까요? 어떤 것을 기대하시는지 미리 이야기하면 더 효과적일 것 같아요.
저는 이 과정을 통해서
1. 어떤 코드가 더 나은 것인지 의논하는 기회
2. Spring의 기능 사용해보기(특히 스프링 시큐리티)
3. 새로운 권한에 대한 문제 생각해보기

이런 것들을 의도하셨다고 생각했어요

MagicalLas commented:
오늘 완전 끝장을 보겠습니다.
다 해버리겟어!!!!

MagicalLas committed:
Remove unused import

MagicalLas committed:
Exclude encryptionService field from JPA

기존에는 encryptionService도 JPA에 저장하는 필드로 등록되었습니다.
하지만, encryptionService는 JPA에 등록할 수 없습니다.
따라서 JPA의 필드에서 제외해주었습니다.

이러한 일이 벌어지는 것은 User클래스가 두가지 역할을 하고있기 때문입니다.

Related Work
  - ee9f03a78266dbaa2a158364532b23a0178760b2

See Also
  - https://docs.jboss.org/hibernate/jpa/2.1/api/javax/persistence/Transient.html

hannut91 commented:
맞습니다 ㅎㅎ 제가 기대했던 것은 강의에서 알려주는 것들을 온전히 학습하는 것이었는데, 다른 것들을 고민하시느라 온전히 학습하시지 못할까 걱정이 됩니다. 물론 다 잘 알고계셔서 쉬운 내용일 것이라 생각이 들지만 걱정이드는 것은 사실입니다.

Lombok이 제공해주는 기능을 사용하지 않고, 즉 다른 누군가가 작성한 라이브러리를 믿지 않고 내가 모두 코드를 작성한다면, 내가 컨트롤 할 수 있는 코드들이 늘어나서 좋지만, 추가적으로 작성해야 할 코드들이 많아집니다. 믿고 사용할 수 있는 라이브러리를 사용하지 못할 때 단점도 생각해봐야 할 것 같아요.

저도 한 때는 JPA를 사용하기 싫어서 모두 JDBC로 코드를 작성한적이 있는데, 결국에는 나 혼자서 모든걸 만들 수 없고 협업을 통해 서비스를 만들어야 하니, 다른 사람들이 이해하기 쉽고, 배우기 쉬운걸로 가야할 것 같더라구요.

MagicalLas commented:
일단 CI를 통과한 상태로 만들어야겠어요.
빨리 안정적인 상태로 돌아가야겠어요.

MagicalLas commented:
> Lombok이 제공해주는 기능을 사용하지 않고, 즉 다른 누군가가 작성한 라이브러리를 믿지 않고 내가 모두 코드를 작성한다면, 내가 컨트롤 할 수 있는 코드들이 늘어나서 좋지만, 추가적으로 작성해야 할 코드들이 많아집니다. 믿고 사용할 수 있는 라이브러리를 사용하지 못할 때 단점도 생각해봐야 할 것 같아요.

라이브러리를 도입할 때 충분한 논의가 있었다면 괜찮은 것 같습니다. 그러나 지금은 프로젝트를 처음 시작하는 코드자체는 주어지기 때문에 어쩔 수 없는 것 같아요. (커밋로그나 PR에 논의한 흔적이 있다면 좋겠지만요)

저는 JPA를 쓰는 것 자체는 매우 좋아하는데, JPA를 Repository로 바로 사용하는건 꺼려지더군요.
이런 트레이드오프가 있다는걸 이해하고 팀과 이야기하여 비교하면서 정할 수 있다면 어떤 방향으로 가던지 좋을 것 같아요.
그리고 팀의 결정을 존중해야겠죠.

어쨋든, 저는 지금 그런 이유에서 이런 이야기를 꺼내보았던것이었어요. 윤석님과 일하게 된다면 윤석님은 어떤 선택을 할지 궁금했거든요!

hannut91 commented:
참 어려운 주제입니다 ㅎㅎ 혼자하면 사실 아무런 고민이 안되는데... 요즘은 내가 선택한 것이 같이 일하는 사람들의 커리어에도 영향이 있을 수 있다고 생각하니 더 걱정이됩니다.

가장 중요한 것은 팀원들과의 합의인 것 같습니다. 그리고 얼마나 책임질 수 있는지가 더 중요한 것 같습니다. 내가 라이브러리를 도입하기로 했을 때 다른 사람들에게 라이브러리를 사용하는 방법을 가르치는 책임이 있습니다. 라이브러리를 도입하지 않기로 했다면, 그 방법을 가르쳐야 합니다. 저는 이것만 지켜지면 괜찮을 것 같아요. 그렇다고 해서 내가 루비를 좋아하고 잘쓰니 루비로 모두 개발하자! 는 힘들겠지만요 ㅎㅎ

MagicalLas committed:
User default construct에서 encryptionService 설정

기존에는 User의 기본 생성자에서는 아무것도 해주지 않았습니다.
하지만, JPA에서는 reflection을 통해서 객체를 생성하기에 기본 생성자에서 encryptionService를 설정해주어야합니다.
따라서, 기본 생성자에서도 encryptionService를 설정해줍니다.

추가적으로 User의 기본 생성자를 테스트하는 코드와 컨테이너를 생성하는 테스트 코드를 작성하였습니다.

MagicalLas committed:
Spring Security적용

기존에는 인터셉터로 토큰을 처리하고 있었다.
하지만, 스프링 시큐리티를 이용해보아야한다.
따라서, 인터셉터 대신 스프링 시큐리티를 사용해서 인증을 처리한다.

MagicalLas committed:
인증된 유저만 유저 변경 & 삭제 가능

기존에는 유저가 아니어도 유저를 지우거나 변경할 수 있었다.
하지만, 그렇게 한다면 다른 사람이 악의적으로 유저를 변경할 가능성이 있다.
따라서 자신이 아니라면 변경할 수 없으며, 유저가 아니라면 삭제를 못하게 변경하였다.

MagicalLas committed:
exclude UserAuthentication from jacoco

MagicalLas committed:
jwt에서 실제 user id를 encode하도록 변경

MagicalLas committed:
Remove logging

MagicalLas committed:
jwt에서 실제 user id를 encode하도록 변경

bd89122badf66e3ea13df4dbf55107f1e6c75549 에서 빠트린 부분이 있어서 추가합니다.

Rleated Work: bd89122badf66e3ea13df4dbf55107f1e6c75549

MagicalLas commented:
휴~ Done!!

MagicalLas commented:
이번주가 너무 챌린징 했네요.

MagicalLas commented:
@hannut91 
7주차 회고를 작성하고 슬랙에 올렸습니다! 추가적으로 해야 할 작업이 있을까요?

https://codesoom-group.slack.com/archives/C015Y4HM1J8/p1615824209054900

hannut91 commented:
이번주는 라스님과 이런저런 라이브러리 도입과 팀원들간의 커뮤니케이션에 대해 고민을 많이 얘기해서 좋았습니다. 그냥 라이브러리 쓰면 되는거 아니냐 보다는 하나하나 설명할 수 있는 사람이 되야겠다는 생각이 들었어요. 라스님 덕분에요 ㅎㅎ

과제 진행 외에 많은 일이 있으셨군요 ㅠ 라스님이 회사에서 어떤 일을 겪고 있는지는 대화를 많이 못했던 것 같아요. 개발에 대한 고민만 얘기한 것 같아서 아쉽네요. 예전에는 달랩 멘토링 월 모임 때 회고를 해서 이런저런 얘기를 들을 수 있었는데, 코드숨에서는 과제 얘기만 하다보니 이런 부분에 대해서 해소가 안된 것 같아요.

그래도 라스님 나름대로 해결책을 찾고 계신 것 같아서 좋지만, 한편으로는 걱정이 됩니다. 라스님이 힘들 때 주변에 힘이 되는 사람들이 많았으면 좋겠습니다. 코드숨에서 달랩멘토링에서 했던 월간 모임을 다시 하면 어떨까 라는 생각이 들었습니다. 감정적인 부분도 굉장히 중요하니까요

고생많으셨습니다 ㅎㅎ
